File: samples/java/src/main/java/com/agents/patterns/resilience/CheckpointResume.java
package com.agents.patterns.resilience;

import org.springframework.ai.chat.client.ChatClient;

import java.util.ArrayList;
import java.util.List;

/**
 * Pattern: Checkpoint and Resume
 * Each step in a long-running agent workflow is checkpointed immediately after
 * completion. On failure, the workflow can resume from the last saved checkpoint
 * rather than restarting from scratch.
 */
public class CheckpointResume {

    public record Checkpoint(String stepName, String result) {}

    /**
     * Executes each step via the agent and records a checkpoint after each success.
     *
     * @param agent the ChatClient agent to execute each step
     * @param steps ordered list of step descriptions
     * @return list of completed checkpoints (one per step)
     */
    public static List<Checkpoint> runWithCheckpoints(ChatClient agent, List<String> steps) {
        List<Checkpoint> checkpoints = new ArrayList<>();

        for (String step : steps) {
            String result = agent.prompt()
                    .user("Execute this step: " + step)
                    .call()
                    .content();
            checkpoints.add(new Checkpoint(step, result));
        }

        return checkpoints;
    }
}

File: samples/csharp/src/AgentPatterns/Resilience/CheckpointResume.cs
using Microsoft.SemanticKernel.ChatCompletion;
namespace AgentPatterns.Resilience;
public record Checkpoint(string StepName, string Result);
public class CheckpointResume {
    public static async Task<List<Checkpoint>> RunWithCheckpoints(IChatCompletionService svc, List<string> steps) {
        var checkpoints = new List<Checkpoint>();
        foreach (var step in steps) {
            var h = new ChatHistory($"Execute step: {step}");
            var result = (await svc.GetChatMessageContentsAsync(h))[0].Content ?? "";
            checkpoints.Add(new Checkpoint(step, result));
            Console.WriteLine($"[Checkpoint] '{step}' saved");
        }
        return checkpoints;
    }
}
